osutil/disks/disks.go

```diff
@@ -0,0 +1,278 @@
+		maj, min, err := parseDeviceMajorMinor(majorMinor)
+		if err != nil {
+			// bad udev output?
+			return nil, fmt.Errorf("cannot find disk for partition %s, bad udev output: %v", partMountPointSource, err)
+		}
+		d.major = maj
+		d.minor = min
+	} else {
+		// the partition is probably a volume or other non-physical disk, so
+		// confirm that DEVTYPE == disk and return the maj/min for it
+		if devType, ok := props["DEVTYPE"]; ok {
+			if devType == "disk" {
+				return d, nil
+			}
+			// unclear what other DEVTYPE's we should support for this
+			return nil, fmt.Errorf("unsupported DEVTYPE %q for mount point source %s", devType, partMountPointSource)
+		}
+
+		return nil, fmt.Errorf("cannot find disk for partition %s, incomplete udev output", partMountPointSource)
+	}
+
+	return d, nil
+}
+
+func (d *disk) FindMatchingPartitionUUID(label string) (string, error) {
+	// if we haven't found the partitions for this disk yet, do that now
+	if d.partitions == nil {
+		d.partitions = make(map[string]string)
+		// step 1. find all devices with a matching major number
+		// step 2. start at the major + minor device for the disk, and iterate over
+		//         all devices that have a partition attribute, starting with the
+		//         device with major same as disk and minor equal to disk minor + 1
+		// step 3. if we hit a device that does not have a partition attribute, then
+		//         we hit another disk, and shall stop searching
+
+		// TODO: are there devices that have structures on them that show up as
+		//       contiguous devices but are _not_ partitions, i.e. some littlekernel
+		//       devices?
+
+		// start with the minor + 1, since the major + minor of the disk we have
+		// itself is not a partition
+		currentMinor := d.minor
+		for {
+			currentMinor++
+			partMajMin := fmt.Sprintf("%d:%d", d.major, currentMinor)
+			props, err := udevProperties(filepath.Join("/dev/block", partMajMin))
+			if err != nil && strings.Contains(err.Error(), "Unknown device") {
+				// the device doesn't exist, we hit the end of the disk
+				break
+			} else if err != nil {
+				// some other error trying to get udev properties, we should fail
+				return "", fmt.Errorf("cannot get udev properties for partition %s: %v", partMajMin, err)
+			}
+
+			if props["DEVTYPE"] != "partition" {
+				// we ran into another disk, break out
+				break
+			}
+
+			label := props["ID_FS_LABEL"]
+			if label == "" {
+				// this partition does not have a filesystem, and thus doesn't have
+				// a filesystem label - this is not fatal, i.e. the bios-boot
+				// partition does not have a filesystem label but it is the first
+				// structure and so we should just skip it
+				continue
+			}
+
+			partuuid := props["ID_PART_ENTRY_UUID"]
+			if partuuid == "" {
+				return "", fmt.Errorf("cannot get udev properties for partition %s, missing udev property \"ID_PART_ENTRY_UUID\"", partMajMin)
+			}
+
+			d.partitions[label] = partuuid
```

Comment:
shouldn't happen but if we have multiple entries with the same label we should pick the first as the old code, so here we should continue if d.partiions[label] != "" already